atl-checker/src/algorithms/global/multithread.rs

```diff
@@ -0,0 +1,484 @@
+use std::cmp::max;
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::fmt::Debug;
+use std::thread;
+use tracing::{span, trace, Level};
+
+use crate::algorithms::global::com::GMessage::{Terminate, Updates};
+use crate::algorithms::global::com::{
+    GBroker, GBrokerManager, GChannelBroker, GChannelBrokerManager, GMessage,
+};
+use crate::algorithms::global::GlobalAlgorithm;
+use crate::edg::{Edge, ExtendedDependencyGraph, Vertex};
+
+pub struct GWorker<B: GBroker<V>, G: ExtendedDependencyGraph<V>, V: Vertex> {
+    edg: G,
+    assignment: HashMap<V, bool>,
+    broker: B,
+}
+
+impl<B: GBroker<V>, G: ExtendedDependencyGraph<V>, V: Vertex> GlobalAlgorithm<G, V>
+    for GWorker<B, G, V>
+{
+    fn edg(&self) -> &G {
+        &self.edg
+    }
+    fn edg_mut(&mut self) -> &mut G {
+        &mut self.edg
+    }
+    fn v0(&self) -> &V {
+        panic!("A GWorker doesnt have a v0 value")
+    }
+    fn v0_mut(&mut self) -> &mut V {
+        panic!("A GWorker doesnt have a v0 value")
+    }
+    fn assignment(&self) -> &HashMap<V, bool> {
+        &self.assignment
+    }
+    fn assignment_mut(&mut self) -> &mut HashMap<V, bool> {
+        &mut self.assignment
+    }
+    fn dist(&self) -> &VecDeque<HashSet<V>> {
+        panic!("A GWorker doesnt have a dist vecdeque")
+    }
+    fn dist_mut(&mut self) -> &mut VecDeque<HashSet<V>> {
+        panic!("A GWorker doesnt have a dist vecdeque")
+    }
+    fn run(&mut self) -> bool {
+        self.run()
+    }
```

Comment:
This looks like recursion. But I assume it goes to the other definition of `run`. Heh